File: app/src/main/java/endroad/birusa/FragmentChat.java
package endroad.birusa;


import android.os.Bundle;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.ImageView;

import com.firebase.ui.database.FirebaseRecyclerAdapter;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.Query;

import java.util.HashMap;
import java.util.Map;

import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import endroad.birusa.model.Message;
import ru.endroad.birusa.R;

public class FragmentChat extends Fragment implements View.OnClickListener {

    RecyclerView recycler;
    FirebaseRecyclerAdapter mAdapter;
    ImageView mSubmitButton;
    EditText mTextField;
    DatabaseReference mRef;
    final static String DB_CHAT = "chat";
    final static int MAX_CHAT_MESSAGE = 100;

    public FragmentChat() {
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        View rootView = inflater.inflate(R.layout.fragment_chat, container, false);

        mSubmitButton = (ImageView) rootView.findViewById(R.id.view_submit);
        mTextField = (EditText) rootView.findViewById(R.id.view_edit_text);
        rootView.findViewById(R.id.view_submit).setOnClickListener(this);

        LinearLayoutManager llm = new LinearLayoutManager(getContext());
        llm.setStackFromEnd(true);
        recycler = (RecyclerView) rootView.findViewById(R.id.recycleList);
        //recycler.setHasFixedSize(true);
        recycler.setLayoutManager(llm);

        mRef = FirebaseDatabase.getInstance().getReference();

        Query query = mRef.child(DB_CHAT).limitToLast(MAX_CHAT_MESSAGE);
        mAdapter = new FirebaseRecyclerAdapter<Message, ViewHolders.MessageHolder>(Message.class, R.layout.item_message, ViewHolders.MessageHolder.class, query) {
            @Override
            public void populateViewHolder(final ViewHolders.MessageHolder vh, final Message message, int position) {
                vh.bindData(getContext(), message, new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        if (view.getId() == R.id.view_card && ((BaseActivity)getActivity()).getUid().compareTo(message.getUid()) != 0)
                            if ((TextUtils.isEmpty(mTextField.getText().toString()))) {
                                mTextField.setText(vh.getAuthor() + ", ");
                                mTextField.setSelection(mTextField.getText().length());
                            }
                    }
                });
            }

            @Override
            protected void onDataChanged() {

                if (recycler.getAdapter().getItemCount() != 0)
                    recycler.smoothScrollToPosition(recycler.getAdapter().getItemCount() - 1);
            }

        };
        recycler.setAdapter(mAdapter);

        return rootView;
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

    }

    @Override
    public void onClick(View view) {
        if (view.getId() == R.id.view_submit) {
            submitMessage();
        }
    }

    private void submitMessage() {

        final String text = mTextField.getText().toString();

        if (TextUtils.isEmpty(text)) {
            mTextField.setError("Введите текст");
            return;
        }

        setEditingEnabled(false);
        //writeNewPost(userId, text);
        //writeNewPost(text);

        final String userId = ((BaseActivity)getActivity()).getUid();
        writeNewPost(userId, text);

    }

    private void setEditingEnabled(boolean enabled) {
        //mTextField.setEnabled(enabled);
        mSubmitButton.setEnabled(enabled);
    }

    // [START write_fan_out]
    private void writeNewPost(String userId, String text) {

        String key = mRef.child(DB_CHAT).push().getKey();
        Message message = new Message(text, userId, BaseActivity.getTime());

        Map<String, Object> values = message.toMap();
        Map<String, Object> childUpdates = new HashMap<>();
        childUpdates.put(DB_CHAT + "/" + key, values);

        mRef.updateChildren(childUpdates);
        mTextField.setText("");
        setEditingEnabled(true);
    }

}

File: feature.chat/src/main/java/ru/endroad/birusa/feature/chat/FragmentChat.kt
package ru.endroad.birusa.feature.chat

import android.view.View
import androidx.recyclerview.widget.LinearLayoutManager
import com.firebase.ui.database.FirebaseRecyclerAdapter
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.FirebaseDatabase
import kotlinx.android.synthetic.main.fragment_chat.*
import ru.endroad.arena.viewlayer.fragment.BaseFragment
import ru.endroad.birusa.feature.chat.ViewHolders.*
import ru.endroad.birusa.feature.chat.model.Message
import java.util.*

class FragmentChat : BaseFragment() {

	override val layout = R.layout.fragment_chat

	private val databaseReference: DatabaseReference = FirebaseDatabase.getInstance().reference
	private val query = databaseReference.child(DB_CHAT).limitToLast(MAX_CHAT_MESSAGE)

	private val mAdapter = object : FirebaseRecyclerAdapter<Message, MessageHolder>(
		Message::class.java,
		R.layout.item_message,
		MessageHolder::class.java,
		query) {
		public override fun populateViewHolder(vh: MessageHolder, message: Message, position: Int) {
			vh.bindData(message, View.OnClickListener { view ->
				if (view.id == R.id.view_card
					&& uid != message.uid
					&& view_edit_text.text.toString().isEmpty()) {
					view_edit_text.setText("${vh.author}, ")
					view_edit_text.setSelection(view_edit_text.text.length)
				}
			})
		}

		override fun onDataChanged() {
			list.adapter?.itemCount?.let { list.smoothScrollToPosition(it - 1) }
		}
	}

	override fun setupViewComponents() {
		view_submit.setOnClickListener { submitMessage() }

		(list.layoutManager as? LinearLayoutManager)?.stackFromEnd = true

		list.adapter = mAdapter
	}

	private fun submitMessage() {
		val text = view_edit_text.text.toString()
		if (text.isEmpty()) {
			view_edit_text.error = "Введите текст"
			return
		}
		setEditingEnabled(false)
		//writeNewPost(userId, text);
//writeNewPost(text);
		val userId = uid
		writeNewPost(userId, text)
	}

	private fun setEditingEnabled(enabled: Boolean) { //mTextField.setEnabled(enabled);
		view_submit.isEnabled = enabled
	}

	// [START write_fan_out]
	private fun writeNewPost(userId: String, text: String) {
		val key = databaseReference.child(DB_CHAT).push().key
		val message = Message(text,
							  userId,
							  System.currentTimeMillis() / 1000L)
		val values = message.toMap()
		val childUpdates: MutableMap<String, Any> = HashMap()
		childUpdates["$DB_CHAT/$key"] = values
		databaseReference.updateChildren(childUpdates)
		view_edit_text.setText("")
		setEditingEnabled(true)
	}

	val uid: String
		get() = FirebaseAuth.getInstance().currentUser?.uid ?: ""

	companion object {
		const val DB_CHAT = "chat"
		const val MAX_CHAT_MESSAGE = 100
	}
}